src/nf/nf_linear2d_layer_submodule.f90

```diff
@@ -0,0 +1,122 @@
+submodule(nf_linear2d_layer) nf_linear2d_layer_submodule
+  use nf_base_layer, only: base_layer
+  implicit none
+contains
+  module function linear2d_layer_cons(&
+      sequence_length, in_features, out_features&
+  ) result(res)
+    integer, intent(in) :: sequence_length, in_features, out_features
+    type(linear2d_layer) :: res
+
+    res % in_features = in_features
+    res % out_features = out_features
+    res % sequence_length = sequence_length
+  end function linear2d_layer_cons
+
+  module subroutine init(self, input_shape)
+    class(linear2d_layer), intent(in out) :: self
+    integer, intent(in) :: input_shape(:)
+
+    allocate(self % output(self % sequence_length, self % out_features))
+    allocate(self % gradient(self % sequence_length, self % in_features))
+
+    allocate(self % weights(self % in_features, self % out_features))
+    self % weights = 0.1
+
+    allocate(self % biases(self % out_features))
```

Comment:
It's here where we should be using `input_shape` to determine `sequence_length` and `in_features`.

`out_features` is already known from the constructor.